### PR TITLE
[Webfonts] Rename to Fonts API

### LINKED ISSUE #42868: Webfonts: Consider renaming "webfont" to "web font" or "font"
Part of https://github.com/WordPress/gutenberg/issues/41479.

## Problem Statement

The terms`webfonts` and `webfont` are misaligned to terminology of well-known authoritative standards.

[As suggested by @desrosj](https://github.com/WordPress/gutenberg/issues/41479#issuecomment-1184564610), this proposal is to consider renaming web font terminology from "webfont" to "web font". In addition, I've expanded this proposal to consider dropping the word "web" and renaming to "font". 

Why rename?

The renaming is to align the feature to well-known authoritative standards for consistency and maintainability.

[As @desrosj noted](https://github.com/WordPress/gutenberg/issues/41479#issuecomment-1184564610), "I know this is a nit, but I think it's important to get it right and be consistent."

As shown below, industry standard refers to a font as "web font" when its font file is served from a web server to the browser. Though technically the API itself is for fonts, meaning web fonts and/or [locally installed fonts via `local()`](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/src#format()).

## Current Choices

- [ ] Don't rename to keep "webfont"
- [ ] Rename to adopt "web font" instead of "webfont"
- [ ] Rename to adopt "font" instead of "webfont"

### Proposal 1: Rename to Web Font

Fonts hosted on a web server are primarily called "web fonts".

- MDN refers to them as ["web fonts"](https://developer.mozilla.org/en-US/docs/Learn/CSS/Styling_text/Web_fonts#web_fonts) 
>Web fonts
>... CSS allows you to specify font files, available on the web, to be downloaded along with your website as it's accessed. This means that any browser supporting this CSS feature can display the fonts you've specifically chosen. Amazing!

- [From Google Fonts glossary](https://fonts.google.com/knowledge/glossary/web_font):
>A web font is any [font](https://fonts.google.com/knowledge/glossary/font) used in a website’s design that isn’t installed by default on the end user’s device—a counterpart to a [system font](https://fonts.google.com/knowledge/glossary/system_font_web_safe_font).

- [The name of the original working group for web fonts](https://www.w3.org/Fonts/WG/)
- [Caniuse calls it @font-face Web fonts](https://caniuse.com/?search=%40font-face)
- [Google Fonts FAQ page says "web fonts"](https://developers.google.com/fonts/faq#can_i_use_css_features_like_text_shadows_and_rotation_with_web_fonts)
- Google 
- [Adobe calls Typekit "Typekit Web Fonts"](https://fonts.adobe.com/) (see the footer link)
- [Microsoft's docs refer to them as "web font"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bg124110(v=vs.85)#web-font-formats)

This rename would result in the following class and function name changes:

| Current name | New name |
| -- | -- |
| WP_Webfonts` | `WP_Web_Fonts` |
| `WP_Webfonts_Provider_Local` | `WP_Web_Fonts_Provider_Local` |
| `WP_Webfonts_Provider` | `WP_Web_Fonts_Provider` |
| `wp_webfonts()` | `wp_web_fonts()` |
| `wp_register_webfonts()` | `wp_register_web_fonts()` |
| `wp_register_webfont()` | `wp_register_web_font()` |
| `wp_enqueue_webfonts()` | `wp_enqueue_web_fonts()` |
| `wp_enqueue_webfont()` | `wp_enqueue_web_font()` |
| `wp_register_webfont_provider()` | `wp_register_web_font_provider()` |
| `wp_get_webfont_providers()` | `wp_get_web_font_providers()` |

as well as all references within code comments.

### Proposal 2: Rename to Font

[The API itself supports the vision of a font library and management](https://github.com/WordPress/gutenberg/issues/41479#font-library-vision). One part of its functionality is to dynamically generate the `@font-face` ruleset.

What is `@font-face`? [Per MDN web docs](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face):
>The @font-face [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS) [at-rule](https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule) specifies a custom font with which to display text; the font can be loaded from either a remote server or a locally-installed font on the user's own computer.

If "web font" means a font file obtained from a web server, then the API's generation of `@font-face` deals with both web and locally installed fonts (on the user's computer or mobile device). 

To be technically accurate, the API then is not limited to only web fonts (font files downloaded to the browser from a web server). Rather it supports "fonts" and the bigger vision of a font library and font management.

This rename would result in the following class and function name changes:

| Current name | New name |
| -- | -- |
| `WP_Webfonts` | `WP_Fonts` |
| `WP_Webfonts_Provider_Local` | `WP_Fonts_Provider_Local` |
| `WP_Webfonts_Provider` | `WP_Fonts_Provider` |
| `wp_webfonts()` | `wp_fonts()` |
| `wp_register_webfonts()` | `wp_register_fonts()` |
| `wp_register_webfont()` | `wp_register_font()` |
| `wp_enqueue_webfonts()` | `wp_enqueue_fonts()` |
| `wp_enqueue_webfont()` | `wp_enqueue_font()` |
| `wp_register_webfont_provider()` | `wp_register_font_provider()` |
| `wp_get_webfont_providers()` | `wp_get_font_providers()` |

as well as all references within code comments.

### PR DESCRIPTION
**Pull request by hellofromtonya**
Closes #42868.

## What?

Renames the API from Webfonts to Fonts API.

## Why?

The renaming is to align the feature to well-known authoritative standards for consistency and maintainability. Why? These standards do not use "webfont" to describe `@font-face` CSS styles. Rather, they use terms such as "font", "custom font", "downloadable font", and "web font".

## How?

1. All publicly facing functionality is deprecated as part of the temporary BC layer. Notices will be thrown to alert developers where to upgrade their code to the new API.
2. "web" is removed from all inline comments, docblocks, and code names (except for the original which are now deprecated).

## Testing Instructions

1. Active TT3
2. Activate the Gutenberg plugin
3. Reset the global styles to the default by:
   - Go to the Site Editor > Global Styles "Styles Beta" UI
   - Click on the 3 vertical dots
   - Select "Reset to defaults"
   - Click on the "Save" button
4. Examine the dynamically generated `@font-face` stylings in the admin area by:
   - Open Dev Tools in your favorite browser
   - Search for `wp-fonts-local` (which is in the `<head>`)
   - Expand the `<style>` element to inspect the CSS
   - The CSS should be:
```html
<style id="wp-fonts-local">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:normal;font-weight:400;font-display:fallback;src:local("DM Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/dm-sans/DMSans-Regular.ttf') format('truetype');}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:400;font-display:fallback;src:local("DM Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/dm-sans/DMSans-Italic.ttf') format('truetype');}
@font-face{font-family:"DM Sans";font-style:normal;font-weight:700;font-display:fallback;src:local("DM Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/dm-sans/DMSans-Bold.ttf') format('truetype');}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:700;font-display:fallback;src:local("DM Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/dm-sans/DMSans-BoldItalic.ttf') format('truetype');}
@font-face{font-family:"IBM Plex Sans";font-style:normal;font-weight:400;font-display:fallback;src:local("IBM Plex Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexSans-Light.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Sans";font-style:italic;font-weight:400;font-display:fallback;src:local("IBM Plex Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexSans-LightItalic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Sans";font-style:normal;font-weight:200;font-display:fallback;src:local("IBM Plex Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexSans-ExtraLight.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Sans";font-style:italic;font-weight:200;font-display:fallback;src:local("IBM Plex Sans"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexSans-ExtraLightItalic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:400;font-display:fallback;src:local("IBM Plex Mono"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexMono-Text.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:italic;font-weight:400;font-display:fallback;src:local("IBM Plex Mono"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexMono-TextItalic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:700;font-display:fallback;src:local("IBM Plex Mono"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexMono-Bold.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:italic;font-weight:700;font-display:fallback;src:local("IBM Plex Mono"), url('/wp-content/themes/twentytwentytwo/assets/fonts/ibm-plex/IBMPlexSans-BoldItalic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:Inter;font-style:normal;font-weight:200 900;font-display:fallback;src:local(Inter), url('/wp-content/themes/twentytwentytwo/assets/fonts/inter/Inter.ttf') format('truetype');font-stretch:normal;}
</style>
```
4. Examine the front-end by:
   - Go to the front-end of the website
   - Open Dev Tools in your favorite browser
   - Search for `wp-fonts-local` 
   - Expand the `<style>` element to inspect the CSS
   - The CSS should be the same as above.
5. Change to the Blue style variation and modify the "text" typography by:
    - Go back to the Global Styles UI in the admin
    - Select "Browser styles"
    - Select the "blue" variation
    - Select "Browse styles" to go back
    - Select "Typography"
    - Select "Text"
    - Change the size to "4"
    - Change the "Appearance" to "Semi Bold Italic"
    - Inspect the CSS as you did in Step 3. It should be the same.
6. Examine the front-end by:
    - Refresh (may have to refresh multiple times until the blue styling renders)
    - Search for `wp-fonts-local`
   - Expand the `<style>` element to inspect the CSS and verify the CSS.

### Testing a plugin

[This gist is a plugin](https://gist.github.com/hellofromtonya/1bbbc9a5354a905278080a9f022c2f60) that uses Jetpack's Google Font Provider package, which is using the original (deprecated) API functionality and web fonts array structure. So it should continue to work and throw deprecation notices.

Instructions of how to install and test are in the plugin's `README.md` file, which [you can get here](https://gist.github.com/hellofromtonya/1bbbc9a5354a905278080a9f022c2f60).

Note: This plugin has a Google Fonts Provider which means it will print `<style id="wp-fonts-jetpack-google-fonts">..</style>` into the `<head>`.

### CODE REVIEW

File: lib/experimental/class-wp-webfonts-provider-local.php [hunk 1]
```diff
@@ -41,6 +44,8 @@ class WP_Webfonts_Provider_Local extends WP_Webfonts_Provider {
 	 * @since 6.1.0
 	 */
 	public function __construct() {
+		deprecated_function( __METHOD__, 'X.X.X', 'WP_Fonts_Provider_Local' );
```
Comment (ironprogrammer) [inline lib/experimental/class-wp-webfonts-provider-local.php]:
```suggestion
		_deprecated_function( __METHOD__, 'X.X.X', 'WP_Fonts_Provider_Local' );
```

File: lib/experimental/class-wp-webfonts-provider.php [hunk 2]
```diff
@@ -14,119 +19,47 @@
...
-	 */
-	protected $id = '';
+abstract class WP_Webfonts_Provider extends WP_Fonts_Provider {
 
 	/**
-	 * Webfonts to be processed.
+	 * Fonts to be processed.
 	 *
 	 * @since X.X.X
+	 * @deprecated X.X.X Use WP_Fonts::$fonts property instead.
```
Comment (ironprogrammer) [inline lib/experimental/class-wp-webfonts-provider.php]:
```suggestion
	 * @deprecated X.X.X Use WP_Fonts_Provider::$fonts property instead.
```

File: lib/experimental/class-wp-webfonts-provider.php [hunk 3]
```diff
@@ -14,119 +19,47 @@
...
 	protected $webfonts = array();
 
 	/**
-	 * Array of Font-face style tag's attribute(s)
-	 * where the key is the attribute name and the
-	 * value is its value.
+	 * Sets this provider's fonts property.
 	 *
 	 * @since X.X.X
+	 * @deprecated X.X.X Use WP_Fonts::set_fonts() instead.
```
Comment (ironprogrammer) [inline lib/experimental/class-wp-webfonts-provider.php]:
```suggestion
	 * @deprecated X.X.X Use WP_Fonts_Provider::set_fonts() instead.
```

File: lib/experimental/class-wp-webfonts.php [hunk 4]
```diff
@@ -22,23 +22,26 @@
...
 	 *
 	 * @param array|string $to_convert The value to convert into a slug. Expected as the web font's array
 	 *                                 or a font-family as a string.
 	 * @return string|false The font slug on success, or false if the font-family cannot be determined.
 	 */
 	public static function get_font_slug( $to_convert ) {
 		$message = is_array( $to_convert )
-			? 'Use WP_Webfonts_Utils::get l_font_family_from_variation() to get the font family from an array and then WP_Webfonts_Utils::convert_font_family_into_handle() to convert the font-family name into a handle'
-			: 'Use WP_Webfonts_Utils::convert_font_family_into_handle() to convert the font-family name into a handle';
+			? 'Use WP_Fonts_Utils::get l_font_family_from_variation() to get the font family from an array and then WP_Fonts_Utils::convert_font_family_into_handle() to convert the font-family name into a handle'
```
Comment (ironprogrammer) [inline lib/experimental/class-wp-webfonts.php]:
```suggestion
			? 'Use WP_Fonts_Utils::get_font_family_from_variation() to get the font family from an array and then WP_Fonts_Utils::convert_font_family_into_handle() to convert the font-family name into a handle'
```

**Review by ironprogrammer — COMMENTED** [review]
Thanks for this, @hellofromtonya! I've left a few notes for review.

**Comment by hellofromtonya** [comment]
Well darn it all, lots has changed since I created this PR. Merge conflicts are being unruly. Going to rebase and reapply the commits. Then force push to get it back in sync. Then will apply the changes @ironprogrammer suggested (thank you Brian).

**Comment by hellofromtonya** [comment]
- Merge conflicts taken care of ✅ 
- Feedback from @ironprogrammer addressed ✅ 
- Tested it locally: works in the Site Editor and frontend ✅ 

@ironprogrammer @aristath It's ready for your review and testing. If possibly, would like to merge this PR this week to unblock the rest of the API's work. Thanks!

**Review by aristath — APPROVED** [review]
Amazing work here! Looks good to me and works as advertised 👍

**Pull request merged**